jupyter_server/services/contents/fileidmanager.py

```diff
@@ -0,0 +1,126 @@
+            "The path of the DB file used by `FileIdManager`. "
+            "Defaults to `jupyter_data_dir()/file_id_manager.db`."
+        ),
+        config=True,
+    )
+
+    def __init__(self, *args, **kwargs):
+        # pass args and kwargs to parent Configurable
+        super().__init__(*args, **kwargs)
+        # initialize connection with db
+        self.con = sqlite3.connect(self.db_path)
+        self.log.debug("Creating File ID tables and indices")
+        self.con.execute(
+            "CREATE TABLE IF NOT EXISTS Files(id INTEGER PRIMARY KEY, path TEXT NOT NULL UNIQUE)"
+        )
+        self.con.execute("CREATE INDEX IF NOT EXISTS ix_Files_path ON FILES (path)")
+        self.con.commit()
+
+    def _normalize_path(self, path):
+        """Normalizes a given file path."""
+        path = os.path.normcase(path)
+        path = os.path.normpath(path)
+        return path
+
+    def index(self, path):
+        """Adds the file path to the Files table, then returns the file ID. If
+        the file is already indexed, the file ID is immediately returned."""
+        path = self._normalize_path(path)
+        existing_id = self.get_id(path)
+        if existing_id is not None:
+            return existing_id
+
+        cursor = self.con.execute("INSERT INTO Files (path) VALUES (?)", (path,))
+        self.con.commit()
+        return cursor.lastrowid
+
+    def get_id(self, path):
+        """Retrieves the file ID associated with a file path. Returns None if
+        the file path has not yet been indexed."""
+        path = self._normalize_path(path)
+        row = self.con.execute("SELECT id FROM Files WHERE path = ?", (path,)).fetchone()
+        self.con.commit()
+        return row[0] if row else None
+
+    def get_path(self, id):
+        """Retrieves the file path associated with a file ID. Returns None if
+        the ID does not exist in the Files table."""
+        row = self.con.execute("SELECT path FROM Files WHERE id = ?", (id,)).fetchone()
+        self.con.commit()
+        return row[0] if row else None
+
+    def move(self, old_path, new_path, recursive=False):
+        """Handles file moves by updating the file path of the associated file
+        ID.  Returns the file ID."""
+        old_path = self._normalize_path(old_path)
+        new_path = self._normalize_path(new_path)
+        self.log.debug(f"Moving file from ${old_path} to ${new_path}")
+
+        if recursive:
+            old_path_glob = os.path.join(old_path, "*")
+            self.con.execute(
+                "UPDATE Files SET path = ? || substr(path, ?) WHERE path GLOB ?",
+                (new_path, len(old_path) + 1, old_path_glob),
+            )
+            self.con.commit()
+
+        id = self.get_id(old_path)
+        if id is None:
+            return self.index(new_path)
+        else:
+            self.con.execute("UPDATE Files SET path = ? WHERE id = ?", (new_path, id))
+            self.con.commit()
+            return id
```

Comment:
In the latest revision, all private methods don't call `commit()`, and this responsibility is always delegated to the parent public method invoking them. I think this makes more sense since it gives public methods ultimate flexibility in how they wish to invoke private methods.